### PR TITLE
Added UserData parameter to Vm, Vmss, VmssVm cmdlets

### PR DESCRIPTION
**Pull request by Sandido**
## Description

Added the UserData Base64 Encoded property to VM and VMSS and VmssVm cmdlets to allow it to be created, updated, and removed. 
We also added functionality to have the New and Update cmdlets check if the UserData string is base64 encoded. If it is not, the cmdlet will modify the value the user provided by Base64 encoding it, then notifying the user of the new value. 

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/966 
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Compute/Compute/Generated/VirtualMachineScaleSet/Config/NewAzureRmVmssConfigCommand.cs [hunk 1]
```diff
@@ -669,6 +686,21 @@ private void Run()
             {
                 vExtendedLocation = new CM.PSExtendedLocation(this.EdgeZone);
             }
+            
+            if (this.IsParameterBound(c => c.UserData))
+            {
+                if (!ValidateBase64EncodedString.validateStringIsBase64Encoded(this.UserData))
```
Comment (Sandido) [inline src/Compute/Compute/Generated/VirtualMachineScaleSet/Config/NewAzureRmVmssConfigCommand.cs:692]: try moving to Validate class

File: src/Compute/Compute/Generated/VirtualMachineScaleSet/VirtualMachineScaleSetCreateOrUpdateMethod.cs [hunk 2]
```diff
@@ -30,19 +30,31 @@
...
-        public const string SimpleParameterSet = "SimpleParameterSet";
+        public const string SimpleParameterSet = "SimpleParameterSet", DefaultParameter = "DefaultParameter";
         public const int vmssFlexibleOrchestrationModeNetworkAPIVersionMinimumInt = 20201101;
         public const string vmssFlexibleOrchestrationModeNetworkAPIVersionMinimum = "2020-11-01";
 
         public override void ExecuteCmdlet()
         {
+            if (this.IsParameterBound(c => c.UserData))
+            {
+                if (!ValidateBase64EncodedString.validateStringIsBase64Encoded(this.UserData))
```
Comment (Sandido) [inline src/Compute/Compute/Generated/VirtualMachineScaleSet/VirtualMachineScaleSetCreateOrUpdateMethod.cs:51]: try moving to Validate class

File: src/Compute/Compute/Generated/VirtualMachineScaleSet/VirtualMachineScaleSetUpdateMethod.cs [hunk 3]
```diff
@@ -31,15 +31,28 @@
...
     [OutputType(typeof(PSVirtualMachineScaleSet))]
     public partial class UpdateAzureRmVmss : ComputeAutomationBaseCmdlet
     {
+        protected const string ExplicitIdentityParameterSet = "ExplicitIdentityParameterSet";
+    
         public override void ExecuteCmdlet()
         {
+            if (this.IsParameterBound(c => c.UserData))
+            {
+                if (!ValidateBase64EncodedString.validateStringIsBase64Encoded(this.UserData))
```
Comment (Sandido) [inline src/Compute/Compute/Generated/VirtualMachineScaleSet/VirtualMachineScaleSetUpdateMethod.cs:49]: try moving to Validate class

File: src/Compute/Compute/Generated/VirtualMachineScaleSetVM/VirtualMachineScaleSetVMUpdateMethod.cs [hunk 4]
```diff
@@ -29,15 +29,31 @@
...
+                               ObjectParameterSet = "ObjectParameter";
+        private InstanceViewTypes UserDataExpand = InstanceViewTypes.UserData;
+
         public override void ExecuteCmdlet()
         {
+            if (this.IsParameterBound(c => c.UserData))
+            {
+                if (!ValidateBase64EncodedString.validateStringIsBase64Encoded(this.UserData))
+                {
+                    this.UserData = ValidateBase64EncodedString.encodeStringToBase64(this.UserData);
```
Comment (Sandido) [inline src/Compute/Compute/Generated/VirtualMachineScaleSetVM/VirtualMachineScaleSetVMUpdateMethod.cs:52]: try moving to Validate class

File: src/Compute/Compute/VirtualMachine/Config/NewAzureVMConfigCommand.cs [hunk 5]
```diff
@@ -216,6 +230,16 @@ public override void ExecuteCmdlet()
                 vm.CapacityReservation = new CapacityReservationProfile();
                 vm.CapacityReservation.CapacityReservationGroup = new SubResource(this.CapacityReservationGroupId);
             }
+	    
+	    if (this.IsParameterBound(c => c.UserData))
+            {
+                if (!ValidateBase64EncodedString.validateStringIsBase64Encoded(this.UserData))
```
Comment (Sandido) [inline src/Compute/Compute/VirtualMachine/Config/NewAzureVMConfigCommand.cs:236]: try moving to Validate class

File: src/Compute/Compute/VirtualMachine/Operation/NewAzureVMCommand.cs [hunk 6]
```diff
@@ -323,9 +323,30 @@ public class NewAzureVMCommand : VirtualMachineBaseCmdlet
...
+            ParameterSetName = DiskFileParameterSet,
+            HelpMessage = "UserData for the VM, which will be base-64 encoded. Customer should not pass any secrets in here.",
+            ValueFromPipeline = true)]
+        public string UserData { get; set; }
 
         public override void ExecuteCmdlet()
         {
+            if (this.IsParameterBound(c => c.UserData))
+            {
+                if (!ValidateBase64EncodedString.validateStringIsBase64Encoded(this.UserData))
```
Comment (Sandido) [inline src/Compute/Compute/VirtualMachine/Operation/NewAzureVMCommand.cs:343]: try moving to Validate class

File: src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs [hunk 7]
```diff
@@ -118,9 +118,35 @@ public class UpdateAzureVMCommand : VirtualMachineBaseCmdlet
...
+            ParameterSetName = ExplicitIdentityParameterSet,
+            HelpMessage = "UserData for the VM, which will be base-64 encoded. Customer should not pass any secrets in here.",
+            ValueFromPipeline = true)]
+        public string UserData { get; set; }
 
         public override void ExecuteCmdlet()
         {
+            if (this.IsParameterBound(c => c.UserData))
+            {
+                if (!ValidateBase64EncodedString.validateStringIsBase64Encoded(this.UserData))
```
Comment (Sandido) [inline src/Compute/Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:143]: try moving to Validate class

**Comment by Sandido** [comment]
Adding Do Not Merge while I investigate the CI failures and verify the functionality is correct with the API team.

File: src/Compute/Compute/Common/ValidateBase64EncodedString.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+using Microsoft.Azure.Management.Compute.Models;
+using Microsoft.Azure.Management.Internal.Resources;
+using Microsoft.Azure.Commands.Common.Authentication;
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+
+namespace Microsoft.Azure.Commands.Compute.Common
+{
+    public static class ValidateBase64EncodedString
+    {
+        public const string UserDataEncodeNotification = "The provided UserData parameter value was not Base-64 encoded. The cmdlet has automatically changed your value and Base64 encoded it. The new UserData value is ";
```
Comment (isra-fel) [inline src/Compute/Compute/Common/ValidateBase64EncodedString.cs:15]: nit: consistency
```suggestion
        public const string UserDataEncodeNotification = "The provided UserData parameter value was not Base64 encoded. The cmdlet has automatically changed your value and Base64 encoded it. The new UserData value is ";
```

File: src/Compute/Compute/Common/ValidateBase64EncodedString.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+using Microsoft.Azure.Commands.Compute.Common;
+using Microsoft.Azure.Commands.ResourceManager.Common;
+using Microsoft.Azure.Management.Compute.Models;
+using Microsoft.Azure.Management.Internal.Resources;
+using Microsoft.Azure.Commands.Common.Authentication;
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+
+namespace Microsoft.Azure.Commands.Compute.Common
+{
+    public static class ValidateBase64EncodedString
```
Comment (isra-fel) [inline src/Compute/Compute/Common/ValidateBase64EncodedString.cs:13]: This class looks very suitable for unit testing

File: src/Compute/Compute/Generated/VirtualMachineScaleSet/Config/NewAzureRmVmssConfigCommand.cs [hunk 10]
```diff
@@ -274,6 +278,12 @@ public partial class NewAzureRmVmssConfigCommand : Microsoft.Azure.Commands.Reso
             HelpMessage = "Id of the capacity reservation Group that is used to allocate.")]
         [ResourceIdCompleter("Microsoft.Compute/capacityReservationGroups")]
         public string CapacityReservationGroupId { get; set; }
+        
+        [Parameter(
+            Mandatory = false,
+            HelpMessage = "UserData for the VM, which will be base-64 encoded. Customer should not pass any secrets in here.",
```
Comment (isra-fel) [inline src/Compute/Compute/Generated/VirtualMachineScaleSet/Config/NewAzureRmVmssConfigCommand.cs:284]: nit: consistency
```suggestion
            HelpMessage = "UserData for the VM, which will be Base64 encoded. Customer should not pass any secrets in here.",
```

File: src/Compute/Compute/Common/ValidateBase64EncodedString.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        public const string UserDataEncodeNotification = "The provided UserData parameter value was not Base-64 encoded. The cmdlet has automatically changed your value and Base64 encoded it. The new UserData value is ";
+        public static bool validateStringIsBase64Encoded(string encodedString)
```
Comment (isra-fel) [inline src/Compute/Compute/Common/ValidateBase64EncodedString.cs:16]: nit: UpperCamelCase for methods
```suggestion
        public static bool ValidateStringIsBase64Encoded(string encodedString)
```

File: src/Compute/Compute/Common/ValidateBase64EncodedString.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+            {
+                return false;
+            }
+            else
+            {
+                return true;
+            }
+        }
+
+        public static string encodeStringToBase64(string toBeEncoded)
```
Comment (isra-fel) [inline src/Compute/Compute/Common/ValidateBase64EncodedString.cs:33]:
```suggestion
        public static string EncodeStringToBase64(string toBeEncoded)
```

**Review by isra-fel — CHANGES_REQUESTED** [review]
Also it would be cool to add (or edit) an example that uses `-UserData` parameter

File: src/Compute/Compute/Common/ValidateBase64EncodedString.cs (hunk 9 above)
Comment (Sandido) [reply src/Compute/Compute/Common/ValidateBase64EncodedString.cs:13]: Added a test for the functionality in this class.

**Review by isra-fel — APPROVED** [review]
Looks good. Thanks

**Pull request merged**